File: skills/actor-pattern/zig-compendium/03_polling.zig
// Example 03: Polling Actor — Timer-Based Background Work
//
// An actor that combines request-response commands with a background poller.
// The poller sleeps between readings (interval measured from end of work,
// avoiding cron-stampede). Demonstrates:
//   - Poller as a plain function (not a struct)
//   - Poller lifecycle owned by run() via an atomic stop flag
//   - Poller sending commands on the actor's channel
//   - std.Thread.sleep for intervals (no timer primitive needed)
//
// Build & run:
//   cd zig-compendium && zig build-exe 03_polling.zig && ./03_polling

const std = @import("std");
const Channel = @import("channel.zig").Channel;
const print = std.debug.print;

const allocator = std.heap.page_allocator;

// --- Command definition ---

const SensorCmd = union(enum) {
    new_reading: struct {
        value: f64,
        reply: *Channel(void),
    },
    latest: struct {
        reply: *Channel(f64),
    },
    history: struct {
        reply: *Channel(HistoryReply),
    },
};

const HistoryReply = struct {
    readings: []const f64,
};

// --- Poller: a plain function, not a struct ---

fn startPoller(
    cmds: *Channel(SensorCmd),
    stop: *std.atomic.Value(bool),
    interval_ns: u64,
) !std.Thread {
    return try std.Thread.spawn(.{}, struct {
        fn run(c: *Channel(SensorCmd), s: *std.atomic.Value(bool), interval: u64) void {
            var rng = std.Random.DefaultPrng.init(@intCast(std.time.nanoTimestamp() & 0xFFFFFFFF));
            const random = rng.random();

            while (!s.load(.acquire)) {
                // Simulate reading a sensor: random temperature 20-30.
                const reading: f64 = 20.0 + @as(f64, @floatFromInt(random.intRangeAtMost(u32, 0, 1000))) / 100.0;

                var reply = Channel(void).init(allocator, 1);
                defer reply.deinit();
                c.send(.{ .new_reading = .{ .value = reading, .reply = &reply } });
                _ = reply.recv();

                // Sleep after work completes (adaptive interval).
                std.Thread.sleep(interval);
            }
        }
    }.run, .{ cmds, stop, interval_ns });
}

// --- Actor: dispatch thread ---

fn sensorActor(cmds: *Channel(SensorCmd)) void {
    var latest: f64 = 0.0;
    var readings = std.ArrayList(f64).initCapacity(allocator, 64) catch unreachable;

    while (cmds.recv()) |cmd| {
        switch (cmd) {
            .new_reading => |c| {
                latest = c.value;
                readings.append(allocator, c.value) catch {};
                c.reply.send({});
            },
            .latest => |c| {
                c.reply.send(latest);
            },
            .history => |c| {
                // Return a copy so the caller can't see mutations.
                const copy = allocator.dupe(f64, readings.items) catch &.{};
                c.reply.send(.{ .readings = copy });
            },
        }
    }
}

// --- Convenience functions ---

fn getLatest(cmds: *Channel(SensorCmd)) f64 {
    var reply = Channel(f64).init(allocator, 1);
    defer reply.deinit();
    cmds.send(.{ .latest = .{ .reply = &reply } });
    return reply.recv() orelse 0.0;
}

fn getHistory(cmds: *Channel(SensorCmd)) []const f64 {
    var reply = Channel(HistoryReply).init(allocator, 1);
    defer reply.deinit();
    cmds.send(.{ .history = .{ .reply = &reply } });
    const r = reply.recv() orelse return &.{};
    return r.readings;
}

// --- Demo ---

pub fn main() !void {
    var cmds = Channel(SensorCmd).init(allocator, 64);
    defer cmds.deinit();

    const actor = try std.Thread.spawn(.{}, sensorActor, .{&cmds});

    // Start the poller: read every 200ms.
    var stop_flag = std.atomic.Value(bool).init(false);
    const poller = try startPoller(&cmds, &stop_flag, 200_000_000);

    // Let the poller collect some readings.
    std.Thread.sleep(1_000_000_000); // 1 second

    const latest = getLatest(&cmds);
    print("Latest reading: {d:.2} C\n", .{latest});

    const history = getHistory(&cmds);
    defer allocator.free(history);
    print("Collected {d} readings:\n", .{history.len});
    for (history, 0..) |r, i| {
        print("  [{d}] {d:.2} C\n", .{ i, r });
    }

    // Shut down: stop the poller, then close the command channel.
    stop_flag.store(true, .release);
    poller.join();
    cmds.close();
    actor.join();
}

File: skills/actor-pattern/rust-compendium/src/bin/03_polling.rs
//! Example 03: Polling Actor -- Timer-Based Background Work
//!
//! An actor that combines request-response commands with a background poller.
//! The poller uses sleep-after-work to avoid the cron-stampede problem.
//! Demonstrates:
//!   - Poller as a plain function spawning a thread
//!   - Poller lifecycle owned by the actor via a stop flag and JoinHandle
//!   - Poller sending commands on the actor's channel
//!   - Sleep after work completes (adaptive interval)
//!   - Shutdown via dropping all sender clones; poller checks its stop flag
//!
//! Run: cargo run --bin 03_polling

use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::mpsc::{self, Sender};
use std::sync::Arc;
use std::thread;
use std::time::Duration;

// --- Command enum ---

enum Command {
    NewReading {
        value: f64,
        reply: Sender<()>,
    },
    Latest {
        reply: Sender<f64>,
    },
    History {
        reply: Sender<Vec<f64>>,
    },
}

// --- Poller ---
// A plain function, not a struct. It has no meaningful state beyond lifecycle
// coordination.

fn start_poller(
    read_sensor: fn() -> f64,
    cmd_tx: Sender<Command>,
    stop: Arc<AtomicBool>,
    interval: Duration,
) -> thread::JoinHandle<()> {
    thread::spawn(move || {
        while !stop.load(Ordering::Relaxed) {
            let reading = read_sensor();

            let (reply_tx, reply_rx) = mpsc::channel();
            // If the actor has shut down, send will fail -- just exit.
            if cmd_tx
                .send(Command::NewReading {
                    value: reading,
                    reply: reply_tx,
                })
                .is_err()
            {
                break;
            }
            // Wait for the actor to process it.
            let _ = reply_rx.recv();

            // Sleep AFTER work completes (adaptive interval).
            // Check stop flag in small increments to allow prompt shutdown.
            let steps = 10;
            let step_dur = interval / steps;
            for _ in 0..steps {
                if stop.load(Ordering::Relaxed) {
                    return;
                }
                thread::sleep(step_dur);
            }
        }
    })
}

// --- Actor handle ---

#[derive(Clone)]
struct SensorService {
    tx: Sender<Command>,
}

impl SensorService {
    fn new(read_sensor: fn() -> f64, poll_interval: Duration) -> Self {
        let (tx, rx) = mpsc::channel::<Command>();
        let poller_tx = tx.clone();

        let stop = Arc::new(AtomicBool::new(false));
        let poller_stop = stop.clone();

        // Start the poller thread.
        let poller_handle = start_poller(read_sensor, poller_tx, poller_stop, poll_interval);

        thread::spawn(move || {
            // All mutable state lives here.
            let mut readings: Vec<f64> = Vec::new();
            let mut latest: f64 = 0.0;

            while let Ok(cmd) = rx.recv() {
                match cmd {
                    Command::NewReading { value, reply } => {
                        latest = value;
                        readings.push(value);
                        let _ = reply.send(());
                    }
                    Command::Latest { reply } => {
                        let _ = reply.send(latest);
                    }
                    Command::History { reply } => {
                        let _ = reply.send(readings.clone());
                    }
                }
            }

            // Actor is shutting down -- stop the poller.
            stop.store(true, Ordering::Relaxed);
            let _ = poller_handle.join();
        });

        SensorService { tx }
    }

    fn latest(&self) -> f64 {
        let (reply_tx, reply_rx) = mpsc::channel();
        self.tx.send(Command::Latest { reply: reply_tx }).unwrap();
        reply_rx.recv().unwrap()
    }

    fn history(&self) -> Vec<f64> {
        let (reply_tx, reply_rx) = mpsc::channel();
        self.tx.send(Command::History { reply: reply_tx }).unwrap();
        reply_rx.recv().unwrap()
    }
}

// --- Simulated sensor ---

fn read_sensor() -> f64 {
    use std::collections::hash_map::DefaultHasher;
    use std::hash::{Hash, Hasher};
    // Pseudo-random based on thread id and time.
    let mut hasher = DefaultHasher::new();
    std::thread::current().id().hash(&mut hasher);
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_nanos()
        .hash(&mut hasher);
    let hash = hasher.finish();
    // Map to 20.0..30.0 range.
    20.0 + (hash % 10000) as f64 / 1000.0
}

// --- Demo ---

fn main() {
    let svc = SensorService::new(read_sensor, Duration::from_millis(200));

    // Let the poller collect some readings.
    thread::sleep(Duration::from_secs(1));

    let latest = svc.latest();
    println!("Latest reading: {:.2}\u{00b0}C", latest);

    let history = svc.history();
    println!("Collected {} readings:", history.len());
    for (i, r) in history.iter().enumerate() {
        println!("  [{}] {:.2}\u{00b0}C", i, r);
    }

    // Dropping `svc` (and the contained Sender) shuts down the actor.
    // The actor thread then signals the poller to stop.
    drop(svc);

    // Give the background threads a moment to clean up.
    thread::sleep(Duration::from_millis(100));
    println!("Shutdown complete.");
}
